DPL: force alignment to 64 bytes

ktf opened the pull request.

ktf commented:
@davidrohr AFAICT, this is all the places we need to touch. In principle this should already work (but I've not verified). On my side, I see no use case in having different alignments, but of course we can try to expose that to the user / allow tuning.

davidrohr commented:
@ktf : There is a use case I see for alignment != 64 bytes. When I use OutputAllocator::make<std::vector<foo>>(...), and the alignment of type foo is larger than 64 bytes, that alignment should be used. I would query that using the C++ alignof keyword.

ktf commented:
The alignment required by C++ is the alignment of the largest member in a struct, unless manually specified, not the padded size of the struct. I can see a reason to cacheline align (i.e. 64bytes, e.g. for AVX,  cache performance) but do we really have use cases where more than that is needed / required?

davidrohr commented:
I think 64bytes will in practice be fine for all data types.
However, the standard defines that allocations and also std::vector<foo> should respect the alignment of foo, which can in principle be larger than 64 bytes. So I would just make this compatible to the behavior one expects from a std::vector.

I mean, does this involve too much work? I was assuming one could just do an alignof(foo) inside DataAllocator::make and propagate the alignment constraint?
If it becomes too complicated, then OK, it is not strictly needed for now, but I would just try to get this right for general principle, because incorrect aligned memory can be a problem that is hard to spot later.

davidrohr commented on DataFormats/MemoryResources/include/MemoryResources/MemoryResources.h:
```diff
@@ -208,7 +208,7 @@ class SpectatorAllocator : public boost::container::pmr::polymorphic_allocator<T
   }
 
-  T* allocate(size_t size) { return reinterpret_cast<T*>(this->resource()->allocate(size * sizeof(T), 0)); }
+  T* allocate(size_t size) { return reinterpret_cast<T*>(this->resource()->allocate(size * sizeof(T), 64)); }
   void deallocate(T* ptr, size_t size)
   {
     this->resource()->deallocate(const_cast<typename std::remove_cv<T>::type*>(ptr), size);
```
There might be one point we have to consider:
C++ uses different allocators for aligned / non-aligned memory.
Therefore, when using new / delete or operator new / delete, one has to pass the alignment constraints also to delete.
This will make sure the correct deallocator is called.
@ktf  @rbx: Are we doing this at the moment?

Normally, this will not break when the mismatched deallocator is used, because by default they are the same.
But e.g. if someone overloads the unaligned new / delete but not the aligned ones, it could theoretically break.

rbx replied:
In FairMQ currently only the shared memory transport honors the alignment argument. And it passes that to boost via `boost::interprocess::managed_shared_memory::allocate_aligned(size, alignment)`. In their docs the corresponding deallocation method to has no alignment argument, nor is one available in the API. So I assume it does the right thing.

davidrohr replied:
OK, but then I guess you should still foresee the alignment parameter in the deallocate function, such that user can pass it already.
This will then allow you to extent aligned alloc/dealloc to other allocators, which require the match.

davidrohr commented:
@ktf @rbx : It seems this is not working.
I made a small test where I just create a couple of messages and print their alignment.
- With the current O2 dev (which doesn't have the shared memory enabled), the minimum alignment seems to be 8 bytes. I guess that comes from ZeroMQ, and as @rbx said, in that case FairMQ is not enforcing any alignment.
- If I enable the shared memory, the minimum alignment increases to 16 bytes, but not 64.
The test was with fairmq 1.4.18.

rbx commented:
Here is our test for this:
https://github.com/FairRootGroup/FairMQ/blob/master/test/message/_message.cxx#L90
So far everything is green.
The call on your side also looks correct. But is the message that you are checking created by one of the modified methods?

davidrohr commented:
`pc.outputs().make<std::vector<char>>(Output{...}, 123)`

rbx commented:
Ok, I think i know what's wrong. The `make` call relies on our FairMQMemoryResource?
I only changed the Message interface so far. The alignment argument in the memory resource is not passed to the transport. Just need to pass it, should be an easy fix.

davidrohr commented:
@rbx @ktf : I have tried with FairMQ 1.4.19 and this patch, but the alignment I get for `pc.outputs().make<std::vector<char>>(...)` is still only 16 bytes.

davidrohr commented:
shall I try again with this? :|

ktf commented:
Yes, albeit I suspect this is not the end of it, as I am not sure of what MemoryResource we are actually using for vectors.

davidrohr commented:
Tried with the new commit, still aligned to 16 bytes.

ktf commented:
@davidrohr this works for me with your testcase.

@matthiasrichter @rbx @mkrzewic: please tell me this is not the way of doing it because I could not find any way of doing something simpler with the current API and frankly this is an abomination. ;-) The only alternative I can think of it that we embed the minAlignment in `fair::mq::ChannelResource` and allow `GetMemoryResource(alignment)` but that would be on the FairMQ side.

mkrzewic commented on Framework/Core/include/Framework/MessageContext.h:
```diff
@@ -145,6 +145,65 @@ class MessageContext
+    }
+
+   private:
+    fair::mq::FairMQMemoryResource* mOther = nullptr; // do you think they'll try to break my balls?
```
// do you think they'll try to break my balls?
probably :), I'm not getting this - why encapsulate here? The MemoryResource API is already ok, so FairMQMemoryResource is suitable so why derive AND encapsulate? I have not see the previous iterations so could be some subtle reason :)

mkrzewic replied:
in the good practice of answering one of my questions: we could indeed use a derived resource to change the default minAlignment so as not to have to specify it in each call, but still not seeing why the encapsulation is there...

ktf replied:
> // do you think they'll try to break my balls?
> probably :),

For the record, the original was:

```
FairMQMemoryResource *mOther; // do you think they'll try to break my balls?
```

which made the Pink Floyd quote much more sophisticated. The `nullptr` spoiled it..

ktf replied:
The problem is that the resource is created by the transport, as a `ChannelResource`, passed as allocator to the vector / container (where I cannot control the alignment) and then passed back to FairMQ when sending (with associated `dynamic_cast` to FairMQResource inside FairMQ). I tried to inherit from ChannelResource, override and use the move constructor from the old resource to the new one, but somehow that does not work, I suspect because FairMQ holds the memory resource so moving it creates issues.

ktf replied:
So basically:
* Encapsulating is needed to override the alignment
* Deriving is needed to send the message

mkrzewic replied:
ah ok, thanks. Then in terms of the "standard" mOther is the upstream resource (or fallback). All perfectly as it should be:)

ktf replied:
yeah, but mUpstream would have not allowed for any pink floyd quote. ;-) Kidding apart, I can rename...

davidrohr commented:
So what is the verdict, do we want to accept this as is, or do we want to move functionality over to FairMQ?

mkrzewic commented:
personally I'd lean towards pushing this as far down as possible - make alignment a parameter of the factory/resource or sth so it can be set once. In simplest case no changes to API, just one new setter. Unless somethign more elaborate is preferred.

rbx commented:
I could add `fair::mq::TransportFactory::SetDefaultAlignment(fair::mq::Alignment alignment)`. That would still be overridable by the message creation methods when necessary.

It would be also accessible through `fair::mq::ChannelResource::getTransportFactory()->SetDefaultAlignment(fair::mq::Alignment alignment)`.

ktf commented:
Yeah, I fixed formatting and changed mOther to mUpstream. I will wait for one test to succeed and then merge. We will let @rbx and @dennisklein to come up with a nicer implementation. Hopefully the virtual call will be optimized out (maybe I should change it to `final`...).

matthiasrichter commented on Framework/Core/include/Framework/MessageContext.h:
```diff
@@ -145,6 +145,65 @@ class MessageContext
 
+  // A memory resource which can force a minimum alignment, so that
+  // the whole polymorphic allocator business is happy...
+  class AlignedMemoryResource : public pmr::FairMQMemoryResource
```
should that go to `DataFormats/MemoryResources` ?

davidrohr commented:
I think ideally it should go to fairmq eventually

ktf commented:
@matthiasrichter yes, we can move it there if @mkrzewic agrees, however I also agree with @davidrohr we should try to move it upstream to FairMQ. @rbx

mkrzewic commented:
@ktf since temporary solutions sometimes like to stick forever maybe it would be good to move it to MemoryResources as @matthiasrichter suggested.

ktf commented:
@mkrzewic then it's probably better we leave it as it is, so that no one uses it... ;-)
